pkg/cmd/bazci/bazci.go

```diff
@@ -66,6 +77,185 @@ tests in Teamcity as painless as possible.`,
+		response, err := s.handleBuildEvent(stream.Context(), in)
+		if err != nil {
+			return err
+		}
+		if err := stream.Send(response); err != nil {
+			return err
+		}
+	}
+	return s.Finalize()
+}
+
+// The returned bool is true iff this is the last event in the stream.
+func (s *monitorBuildServer) handleBuildEvent(
+	ctx context.Context, in *build.PublishBuildToolEventStreamRequest,
+) (*build.PublishBuildToolEventStreamResponse, error) {
+	switch event := in.OrderedBuildEvent.Event.Event.(type) {
+	case *build.BuildEvent_BazelEvent:
+		var bazelBuildEvent bes.BuildEvent
+		any := &ptypes.Any{
+			TypeUrl: event.BazelEvent.TypeUrl,
+			Value:   event.BazelEvent.Value,
+		}
+		if err := ptypes.UnmarshalAny(any, &bazelBuildEvent); err != nil {
+			return nil, err
+		}
+		switch id := bazelBuildEvent.Id.Id.(type) {
+		case *bes.BuildEventId_NamedSet:
+			namedSetID := id.NamedSet.Id
+			namedSet := bazelBuildEvent.GetNamedSetOfFiles()
+			var files []builtArtifact
+			for _, file := range namedSet.Files {
+				uri := file.GetUri()
+				files = append(files, builtArtifact{src: strings.TrimPrefix(uri, "file://"), dst: file.Name})
+			}
+			for _, set := range namedSet.FileSets {
+				files = append(files, s.namedSetsOfFiles[set.Id]...)
+			}
+			s.namedSetsOfFiles[namedSetID] = files
+		case *bes.BuildEventId_TestResult:
+			res := fullTestResult{
+				run:        id.TestResult.Run,
+				shard:      id.TestResult.Shard,
+				attempt:    id.TestResult.Attempt,
+				testResult: bazelBuildEvent.GetTestResult(),
+			}
+			s.testResults[id.TestResult.Label] = append(s.testResults[id.TestResult.Label], res)
+		case *bes.BuildEventId_TargetCompleted:
+			s.builtTargets[id.TargetCompleted.Label] = bazelBuildEvent.GetCompleted()
+		case *bes.BuildEventId_TestSummary:
+			label := id.TestSummary.Label
+			outputDir := strings.TrimPrefix(label, "//")
+			outputDir = strings.ReplaceAll(outputDir, ":", "/")
+			outputDir = filepath.Join("bazel-testlogs", outputDir)
+			summary := bazelBuildEvent.GetTestSummary()
+			for _, testResult := range s.testResults[label] {
+				outputDir := outputDir
+				if testResult.run > 1 {
+					outputDir = filepath.Join(outputDir, fmt.Sprintf("run_%d", testResult.run))
+				}
+				if summary.ShardCount > 1 {
+					outputDir = filepath.Join(outputDir, fmt.Sprintf("shard_%d_of_%d", testResult.shard, summary.ShardCount))
+				}
+				if testResult.attempt > 1 {
+					outputDir = filepath.Join(outputDir, fmt.Sprintf("attempt_%d", testResult.attempt))
+				}
+				for _, output := range testResult.testResult.TestActionOutput {
+					if output.Name == "test.log" || output.Name == "test.xml" {
+						src := strings.TrimPrefix(output.GetUri(), "file://")
+						dst := filepath.Join(artifactsDir, outputDir, filepath.Base(src))
+						if err := doCopy(src, dst); err != nil {
+							return nil, err
+						}
+					} else {
+						panic(output)
+					}
+				}
+			}
+		}
+		// TODO: anything to do in this case?
+		//if event.LastMessage {}
```

Comment:
Maybe we should use it instead of `BuildFinished` in-case we decide to use events after `BuildFinished` in the future?

Reply:
Turns out that `LastMessage` is not necessary for us in this case. I'm just removing the TODO.